### PR TITLE
Compile string literals as stdlib Strings

### PR DESCRIPTION
**Pull request by segiddins**
- [x] Update stdlib
- [x] Update examples
- [x] Ship the stdlib inside the runtime framework
- [x] Make detecting stdlib presence more robust
- [x] Allow string literals to be coerced to `*Int`s?

### CODE REVIEW

**Comment by harlanhaskins** [comment]
I'd say definitely -- if we can demote a `String` to `*Int8` automatically it'll help interface with C tremendously.

File: Sources/IRGen/ValueIRGen.swift [hunk 1]
```diff
@@ -152,7 +153,17 @@ extension IRGenerator {
   }
   
   func visitStringExpr(_ expr: StringExpr) -> Result {
-    return codegenGlobalStringPtr(expr.value)
+    let globalString = codegenGlobalStringPtr(expr.value)
+    let zero = IntType.int64.zero()
+    let indices = [zero, zero]
+    let ptr = globalString.ptr.constGEP(indices: indices)
```
Comment (harlanhaskins) [inline Sources/IRGen/ValueIRGen.swift]: I think this can just be a `load` no?

File: Sources/IRGen/ValueIRGen.swift [hunk 2]
```diff
@@ -152,7 +153,17 @@ extension IRGenerator {
...
+    guard let stringTypeDecl = context.type(named: "String") else { fatalError("use of string literal without stdlib String") }
```
Comment (harlanhaskins) [inline Sources/IRGen/ValueIRGen.swift]: Maybe we should make this a property of `ASTContext` like we discussed

File: Sources/IRGen/ValueIRGen.swift [hunk 3]
```diff
@@ -152,7 +153,17 @@ extension IRGenerator {
...
+    guard let inititalizer = stringTypeDecl.initializers.first(where: { initializer in
```
Comment (harlanhaskins) [inline Sources/IRGen/ValueIRGen.swift]: I'd prefer not to use string comparison for this. It can get sloppy.

**Review by harlanhaskins — CHANGES_REQUESTED** [review]

File: stdlib/String.tr [hunk 4]
```diff
@@ -122,3 +128,7 @@ func +(lhs: *Int8, rhs: *Int8) -> String {
   str.append(rhs)
   return str
 }
+
+func trill_fatalError(_ message: String) {
```
Comment (harlanhaskins) [inline stdlib/String.tr]: I think this should be spelled `fatalError(_:)`

File: stdlib/bytearray.tr [hunk 5]
```diff
@@ -51,7 +50,7 @@ indirect type ByteArray {
   }
   func _boundsCheck(_ index: Int) {
     if index > self.length {
-      trill_fatalError("index out of bounds")
+      trill_fatalError("index out of bounds".cString())
```
Comment (harlanhaskins) [inline stdlib/bytearray.tr]: Probably don't need this now

File: Sources/Semantic Analysis/Sema.swift [hunk 6]
```diff
@@ -247,31 +247,6 @@ class Sema: ASTTransformer, Pass {
     return true
   }
   
-  func candidate(forArgs args: [Argument], candidates: [FuncDecl]) -> FuncDecl? {
```
Comment (harlanhaskins) [inline Sources/Semantic Analysis/Sema.swift:248]: Ha! Probably should've removed this weeks ago.

File: stdlib/String.tr (hunk 4 above)
Comment (harlanhaskins) [reply stdlib/String.tr]: When we get default arguments, this will be fun.

**Comment by harlanhaskins** [comment]
This will need updating once we merge #36

**Comment by segiddins** [comment]
Still uses a string compare, but other than that using the stdlib is now much more robust!

**Comment by segiddins** [comment]
A _super_ nice side effect to this is that `-emit ast` no longer includes the stdlib unless you add `-show-imports`!

**Comment by harlanhaskins** [comment]
🎉 Why is that?

**Comment by segiddins** [comment]
> Why is that?

Because you no longer explicitly pass in the stdlib files to be compiled

**Comment by harlanhaskins** [comment]
Oh! Right! 😃

File: Sources/Base/ASTContext.swift [hunk 7]
```diff
@@ -338,6 +339,15 @@ public class ASTContext {
         expr.type = contextualType
         return true
       }
+    case let expr as StringExpr:
+      if case .pointer(type: DataType.int8) = canTy {
```
Comment (harlanhaskins) [inline Sources/Base/ASTContext.swift]: `DataType` is `Equatable` so you can rewrite this as:

```swift
if [.string, .pointer(type: .int8)].contains(canTy) {
  expr.type = contextualType
  return true
}
```

File: Sources/Infrastructure/Utilities.swift [hunk 8]
```diff
@@ -72,3 +72,14 @@ let ansiEscapeSupportedOnStdErr: Bool = {
     
     return true
 }()
+
+let runtimeFramework: Bundle? = Bundle(identifier: "com.harlanhaskins.trillRuntime")
```
Comment (harlanhaskins) [inline Sources/Infrastructure/Utilities.swift]: We might want to rename this to `com.trill-lang.trillRuntime`

File: Sources/LLVM Wrappers/LLVMWrappers.cpp [hunk 9]
```diff
@@ -88,25 +90,28 @@ RawOptions ParseArguments(int argc, char **argv) {
...
-  cl::opt<bool> onlyDiagnostics("diagnostics-only", cl::desc("Only emit diagnostics"));
-  cl::opt<std::string> target("target", cl::desc("Override the LLVM target machine"));
-  cl::opt<std::string> outputFile("o", cl::desc("Output filename"));
-  cl::list<std::string> filenames(cl::Positional, cl::desc("<filenames>"));
+                                           clEnumValEnd), cat);
+  cl::opt<bool> jit("run", cl::desc("JIT the specified files"), cat);
+  cl::opt<bool> parseOnly("parse-only", cl::desc("Only parse, do not run semantic analysis"), cat);
+  cl::opt<bool> showImports("show-imports", cl::desc("Show imported items in the AST dump"), cat);
+  cl::opt<bool> stdlib("stdlib", cl::desc("Include the trill standard library"), cat);
+  stdlib = true;
```
Comment (harlanhaskins) [inline Sources/LLVM Wrappers/LLVMWrappers.cpp:98]: This should be flipped, no?

```c++
cl::opt<bool> noStdlib("no-stdlib", cl::desc("Don't include the trill standard library"), cat);
```

File: Sources/LLVM Wrappers/LLVMWrappers.cpp [hunk 10]
```diff
@@ -144,6 +149,7 @@ RawOptions ParseArguments(int argc, char **argv) {
     isStdin,
     jsonDiagnostics,
     showImports,
+    static_cast<bool>(stdlib),
```
Comment (harlanhaskins) [inline Sources/LLVM Wrappers/LLVMWrappers.cpp]: No need to `static_cast` here, and if this is flipped we should just `!` it.

File: Sources/trill/main.swift [hunk 11]
```diff
@@ -48,6 +34,24 @@ func populate(driver: Driver, options: Options,
     gen = irgen
   }
   
+  if options.includeStdlib {
+    driver.add("Lexing and Parsing the trill standard library") { context in
```
Comment (harlanhaskins) [inline Sources/trill/main.swift]: Pass title is kinda long. Maybe just `Parsing Standard Library`

**Review by harlanhaskins — CHANGES_REQUESTED** [review]
I'm ridiculously excited to merge this.

File: Sources/LLVM Wrappers/LLVMWrappers.cpp (hunk 9 above)
Comment (segiddins) [reply Sources/LLVM Wrappers/LLVMWrappers.cpp:98]: I don't entirely understand the format used by the default bool parser, so I'm not sure -- we want both `-stdlib` and `-no-stdlib` to work (and not require `-no-no-stdlib`).

Comment (harlanhaskins) [reply Sources/LLVM Wrappers/LLVMWrappers.cpp:98]: `cl::opt` will automatically marshall to and from `bool`

What would we need `-stdlib` for? That should be the default.

Comment (segiddins) [reply Sources/LLVM Wrappers/LLVMWrappers.cpp:98]: It is the default, but you still want it so you can override an earlier `-no-stdlib`

Comment (harlanhaskins) [reply Sources/LLVM Wrappers/LLVMWrappers.cpp:98]: I don't necessarily think that's a priority for trill 😅 I don't know of anybody who's generating build flags for trill files...

Comment (harlanhaskins) [reply Sources/LLVM Wrappers/LLVMWrappers.cpp:98]: But if you'd prefer to have a `no-stdlib` flag, then that's fine.

File: Sources/LLVM Wrappers/LLVMWrappers.cpp [hunk 12]
```diff
@@ -88,25 +90,28 @@ RawOptions ParseArguments(int argc, char **argv) {
...
+                                    cl::desc("<extra linker flags>"), cat);
   cl::list<std::string> ccFlags("Xcc", cl::Positional,
                                 cl::PositionalEatsArgs,
-                                cl::desc("<extra clang flags>"));
+                                cl::desc("<extra clang flags>"), cat);
   cl::list<std::string> jitArgs("args", cl::Positional,
                                 cl::PositionalEatsArgs,
-                                cl::desc("<JIT arguments>"));
+                                cl::desc("<JIT arguments>"), cat);
+  cl::HideUnrelatedOptions(category);
```
Comment (segiddins) [inline Sources/LLVM Wrappers/LLVMWrappers.cpp:114]: this means `trill -help` will only print our flags, and not all the ones inherited from LLVM

Comment (harlanhaskins) [reply Sources/LLVM Wrappers/LLVMWrappers.cpp:114]: Yep, which is great!

**Review by harlanhaskins — APPROVED** [review]

**Comment by harlanhaskins** [comment]
:shipit:!!!!!!!!!!!!! (once CI passes)

**Comment by segiddins** [comment]
🚢

**Pull request merged**